File: src/platforms/web/src/api/trade.ts
import request from '@/api/request'

export interface TradeAccount {
    id: number
    name: string
    broker: string
    cash_balance: number
    total_assets: number
    market_value: number
    position_ratio_percent: number
    today_pnl: number
    unrealized_pnl: number
    accumulated_pnl: number
    commission_rate: number
    min_commission: number
    stamp_duty_rate: number
    transfer_fee_rate: number
    notes?: string
    positions_count: number
    created_at?: string
}

export interface TradePosition {
    id: number
    account_id: number
    stock_code: string
    stock_name: string
    quantity: number
    cost_price: number
    accumulated_pnl: number
    current_price: number
    change: number
    percent: number
    market_value: number
    cost_amount: number
    unrealized_pnl: number
    unrealized_pnl_percent: number
    today_pnl: number
    weight_percent?: number
    updated_at?: string
}

export interface AssetDistributionItem {
    label: string
    category: 'cash' | 'stock'
    amount: number
    ratio_percent: number
}

export interface PortfolioSummary {
    total_assets: number
    total_market_value: number
    total_cash: number
    total_today_pnl: number
    total_unrealized_pnl: number
    total_accumulated_pnl: number
    position_ratio_percent: number
    accounts_count: number
    positions_count: number
    asset_distribution: AssetDistributionItem[]
}

export interface TradeOrder {
    id: number
    account_id: number
    stock_code: string
    stock_name: string
    trade_type: 'buy' | 'sell'
    price: number
    quantity: number
    amount: number
    commission: number
    stamp_duty: number
    transfer_fee: number
    total_fee: number
    net_amount: number
    realized_pnl: number
    cost_price_after: number
    holding_after: number
    notes?: string
    transacted_at?: string
}

export interface TradeCashFlow {
    id: number
    flow_type: 'deposit' | 'withdraw' | 'trade_buy' | 'trade_sell' | 'dividend'
    amount: number
    balance_after: number
    notes?: string
    transacted_at?: string
}

export interface SimulationResult {
    stock_code: string
    stock_name: string
    price: number
    simulated_quantity: number
    simulated_lots: number
    trade_amount: number
    fees: {
        commission: number
        stamp_duty: number
        transfer_fee: number
        total_fee: number
    }
    total_required: number
    current_cash: number
    is_sufficient: boolean
    cash_shortage: number
    diluted_cost_price: number
    new_quantity: number
    new_stock_market_value: number
    new_total_assets: number
    position_ratio_percent: number
    transfer_suggestions: Array<{
        account_id: number
        name: string
        available_cash: number
        can_cover_fully: boolean
    }>
}

export const getTradeAccounts = () => request.get<TradeAccount[]>('/trade/accounts')
export const createTradeAccount = (data: Partial<TradeAccount> & { initial_cash?: number }) =>
    request.post<{ success: boolean; id: number }>('/trade/accounts', data)
export const updateTradeAccount = (id: number, data: Partial<TradeAccount>) =>
    request.put<{ success: boolean }>('/trade/accounts/' + id, data)
export const deleteTradeAccount = (id: number) =>
    request.delete<{ success: boolean }>('/trade/accounts/' + id)

export const addCashFlow = (accountId: number, data: { flow_type: 'deposit' | 'withdraw'; amount: number; notes?: string }) =>
    request.post<{ success: boolean; flow_id: number; balance_after: number }>('/trade/accounts/' + accountId + '/cash-flow', data)
export const getCashFlows = (accountId: number) =>
    request.get<TradeCashFlow[]>('/trade/accounts/' + accountId + '/cash-flows')

export const getAccountPositions = (accountId: number) =>
    request.get<TradePosition[]>('/trade/accounts/' + accountId + '/positions')

export const getPortfolioSummary = () =>
    request.get<PortfolioSummary>('/trade/summary')

export const createTradeOrder = (data: {
    account_id: number
    stock_code: string
    stock_name: string
    trade_type: 'buy' | 'sell'
    price: number
    quantity: number
    custom_commission?: number
    custom_stamp_duty?: number
    custom_transfer_fee?: number
    notes?: string
}) => request.post<{
    success: boolean
    order_id: number
    trade_type: string
    net_amount: number
    total_fee: number
    realized_pnl: number
    holding_after: number
    cost_price_after: number
}>('/trade/orders', data)

export const getTradeOrders = (params?: { account_id?: number; stock_code?: string; limit?: number }) =>
    request.get<TradeOrder[]>('/trade/orders', { params })

export const simulateOrder = (data: {
    account_id: number
    stock_code: string
    stock_name: string
    price: number
    target_amount?: number
    quantity?: number
}) => request.post<SimulationResult>('/trade/simulate', data)

File: src/api/models/trade.py
from datetime import datetime
from sqlalchemy import (
    Boolean,
    DateTime,
    ForeignKey,
    Numeric,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from api.core.database import Base


class TradeAccount(Base):
    """交易账户模型"""

    __tablename__ = "trade_accounts"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    broker: Mapped[str] = mapped_column(String(50), default="通用券商", nullable=False)
    cash_balance: Mapped[float] = mapped_column(
        Numeric(14, 2), default=0.0, nullable=False
    )
    commission_rate: Mapped[float] = mapped_column(
        Numeric(8, 6), default=0.00025, nullable=False
    )  # 佣金率（默认万2.5）
    min_commission: Mapped[float] = mapped_column(
        Numeric(8, 2), default=5.0, nullable=False
    )  # 最低佣金（默认5元）
    stamp_duty_rate: Mapped[float] = mapped_column(
        Numeric(8, 6), default=0.0005, nullable=False
    )  # 印花税率（默认万5，仅卖出）
    transfer_fee_rate: Mapped[float] = mapped_column(
        Numeric(8, 6), default=0.00001, nullable=False
    )  # 过户费率（默认十万分之1）
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )

    positions: Mapped[list["TradePosition"]] = relationship(
        "TradePosition", back_populates="account", cascade="all, delete-orphan"
    )
    orders: Mapped[list["TradeOrder"]] = relationship(
        "TradeOrder", back_populates="account", cascade="all, delete-orphan"
    )
    cash_flows: Mapped[list["TradeCashFlow"]] = relationship(
        "TradeCashFlow", back_populates="account", cascade="all, delete-orphan"
    )


class TradePosition(Base):
    """账户持仓标的模型"""

    __tablename__ = "trade_positions"
    __table_args__ = (
        UniqueConstraint(
            "account_id",
            "stock_code",
            name="uq_trade_position_account_stock",
        ),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    account_id: Mapped[int] = mapped_column(
        ForeignKey("trade_accounts.id", ondelete="CASCADE"), nullable=False, index=True
    )
    stock_code: Mapped[str] = mapped_column(String(32), nullable=False, index=True)
    stock_name: Mapped[str] = mapped_column(String(64), nullable=False)
    quantity: Mapped[float] = mapped_column(Numeric(14, 2), default=0.0, nullable=False)
    cost_price: Mapped[float] = mapped_column(
        Numeric(14, 4), default=0.0, nullable=False
    )
    accumulated_pnl: Mapped[float] = mapped_column(
        Numeric(14, 2), default=0.0, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )

    account: Mapped["TradeAccount"] = relationship("TradeAccount", back_populates="positions")


class TradeOrder(Base):
    """交易订单流水模型（买入/卖出交割记录）"""

    __tablename__ = "trade_orders"

    id: Mapped[int] = mapped_column(primary_key=True)
    account_id: Mapped[int] = mapped_column(
        ForeignKey("trade_accounts.id", ondelete="CASCADE"), nullable=False, index=True
    )
    stock_code: Mapped[str] = mapped_column(String(32), nullable=False, index=True)
    stock_name: Mapped[str] = mapped_column(String(64), nullable=False)
    trade_type: Mapped[str] = mapped_column(
        String(16), nullable=False
    )  # 'buy' or 'sell'
    price: Mapped[float] = mapped_column(Numeric(14, 4), nullable=False)
    quantity: Mapped[float] = mapped_column(Numeric(14, 2), nullable=False)
    amount: Mapped[float] = mapped_column(Numeric(14, 2), nullable=False)  # price * quantity
    commission: Mapped[float] = mapped_column(
        Numeric(10, 2), default=0.0, nullable=False
    )
    stamp_duty: Mapped[float] = mapped_column(
        Numeric(10, 2), default=0.0, nullable=False
    )
    transfer_fee: Mapped[float] = mapped_column(
        Numeric(10, 2), default=0.0, nullable=False
    )
    total_fee: Mapped[float] = mapped_column(
        Numeric(10, 2), default=0.0, nullable=False
    )
    net_amount: Mapped[float] = mapped_column(
        Numeric(14, 2), nullable=False
    )  # 买入: amount + total_fee; 卖出: amount - total_fee
    realized_pnl: Mapped[float] = mapped_column(
        Numeric(14, 2), default=0.0, nullable=False
    )
    cost_price_after: Mapped[float] = mapped_column(
        Numeric(14, 4), default=0.0, nullable=False
    )
    holding_after: Mapped[float] = mapped_column(
        Numeric(14, 2), default=0.0, nullable=False
    )
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    transacted_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, nullable=False, index=True
    )

    account: Mapped["TradeAccount"] = relationship("TradeAccount", back_populates="orders")


class TradeCashFlow(Base):
    """资金流水模型（银证转账、交易出入款）"""

    __tablename__ = "trade_cash_flows"

    id: Mapped[int] = mapped_column(primary_key=True)
    account_id: Mapped[int] = mapped_column(
        ForeignKey("trade_accounts.id", ondelete="CASCADE"), nullable=False, index=True
    )
    flow_type: Mapped[str] = mapped_column(
        String(32), nullable=False
    )  # 'deposit', 'withdraw', 'trade_buy', 'trade_sell', 'dividend'
    amount: Mapped[float] = mapped_column(
        Numeric(14, 2), nullable=False
    )  # 变动金额（流入为正，流出为负）
    balance_after: Mapped[float] = mapped_column(Numeric(14, 2), nullable=False)
    related_order_id: Mapped[int | None] = mapped_column(
        ForeignKey("trade_orders.id", ondelete="SET NULL"), nullable=True
    )
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    transacted_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, nullable=False, index=True
    )

    account: Mapped["TradeAccount"] = relationship("TradeAccount", back_populates="cash_flows")
